p4-16/psa/PSA.mdk

```diff
@@ -245,76 +245,185 @@ Operations in the egress pipeline will not execute if drop is called in ingress.
+~ Figure { #fig-clone; caption: "Clone/recirculate/resubmit in PSA"; page-align: here; }
+![clone]
+~
+[clone]: clone-recirc-resubmit.pdf { width: 90%; }
 
-Parameters:
+Figure [#fig-clone] show two proposed architectures for clone/recirculate/resubmit in PSA.
 
-- clone_method  The type of cloning.
-- port          The port to send the cloned packet to.
+### PRE Metadata
 
 ```
-void clone     (in CloneMethod_t clone_method, in PortId_t port);
+struct psa_ingress_metadata_for_parser_t {
+   bool is_recirc;
+}
 ```
 
-#### Clone with extra data operation { #sec-pre-clone-with-data }
-
-Create a copy of the packet with additional data and send it to the
-specified port.
-
-The PSA specifies four types of cloning, with the packet sourced
-from different points in the pipeline and sent back to ingress
-or to the buffering queue in the egress (see [#sec-cloning-methods]).
-\TODO: needs both source and destination data
-
-Parameters:
-
-- clone_method  The type of cloning.
-- port          The port to send the cloned packet to.
-- data additional header data attached to the packet
-
 ```
-void clone<T>  (in CloneMethod_t clone_method, in PortId_t port, in T data);
+struct psa_ingress_metadata_for_pre_t {
+   bool clone_ena;
+   CloneSpectT clone_spec;
+   bool resubmit_ena;
+}
 ```
 
-#### Resubmit operation
-
-Send a packet to the ingress port with additional data appended.
-
-This operation is intended for recursive packet processing.
-\TODO: needs both source and destination data
-
-Parameters:
-
-- data A header definition that can be added to the set of
-  packet headers.
-- port The input port at which the packet will be resubmitted.
-
 ```
-void resubmit<T>(in T data, in PortId_t port);
+struct psa_egress_metadata_for_pre_t {
+   bool clone_ena;
+   CloneSpecT clone_spec;
+   bool recirc_ena;
+}
 ```
 
-#### Recirculate operation
-
-Send a post deparse packet to the ingress port with additional data
-appended.
-
-This operation is intended for recursive packet processing.
-\TODO: needs both source and destination data
-
-Parameters:
-
-- data A header definition that can be added to the set of packet headers.
-- port The input port at which the packet will be resubmitted.
+#### Clone
+
+A PSA implementation provides a clone mechanism to create a copy of
+the original packet as an independent packet instance. The cloned
+packet can be a duplicate of the original packet or a copy of the
+deparsed packet after egress pipeline. The clone mechanism can submit
+the cloned packet to ingress parser or buffering mechanism. In a PSA
+implementation, the clone mechanism is implemented in PRE.
+
+The clone mechanism can optionally attach metadata to the cloned
+packet. A PSA implementation provides the `clone` extern to specify
+the attached metadata. The `clone` extern provides a `emit` method
+which accepts the attached metadata of generic type `T`. In a PSA
+implementation, the metadata is prepended to the cloned packet. It is
```

Comment:
... in other way, it will show up in packet_in extern representing the cloned (but not the original) packet.